src/zmqsubscriber/zmqsubscriber.go

```diff
@@ -109,14 +121,27 @@ func (z *ZMQSubscriber) Run() error {
 
 		topic, payload := string(msg[0]), msg[1:]
 		log.Printf(`[ZMQ] received topic "%s"`, topic)
-		switch topic {
-		case TOPIC_HASHTX:
-			z.IncomingTx <- parseTransaction(t, payload)
-		case TOPIC_RAWBLOCK:
-			z.IncomingBlocks <- parseBlock(t, payload)
-		default:
-			return fmt.Errorf("unknown topic %s", topic)
-		}
+
+
+		// process received messages asynchronously so that we do not stall the queue while parsing
+		go func() {
+			switch topic {
```

Comment:
could be moved into `if err := z.processMessage(payload); err != nil { parseErrors <- err }` to reduce boilerplate

Reply:
or even a `z.handleZMQHashTx(firstSeen time, payload string, parseError chan err)` method which  does the handling

Reply:
`processMessage(payload) error` would have a slightly more standard interface

Reply:
Oh you mean the whole _anonymous inline function_, correct? That's what I meant here: https://github.com/0xB10C/bademeister-go/pull/16#discussion_r340013756. Additionally a `handleTx` and handleBlock function ().

basically:

```go

go handleZMQMessage(topic, payload, time, parseErrors)

...

func handleZMQMessage(...) {

	switch topic {
		case TOPIC_HASHTX:
			z.handleZMQHashTxMessage(topic, payload, time, parseErrors)
		}
		case TOPIC_RAWBLOCK:
			z.handleZMQRawBlockMessage(topic, payload, time, parseErrors)	
		}
	    default:
		....
	}

} 
```

Reply:
oops, I meant to :+1: the other comment (only `handleZMQMessage(topic, payload) error`)

I don't think we need `handleZMQHashTxMessage()` since we already have `parseTransaction(time, []byte) error`. 

Funcs that have an error as the return value are easier to read and test than funcs that put errors on a channel, so I would keep that part in the anonymous inline func (as it is an implementation detail of the loop).